src/int/specialized_div_rem/trifecta.rs

```diff
@@ -0,0 +1,451 @@
+                    let extra_shl = duo_extra - div_extra;
+
+                    // Addition to the quotient.
+                    quo += (quo_part << extra_shl);
+
+                    // Subtraction from `duo`. At least `n_h - 1` bits are cleared from `duo` here.
+                    duo -= (div.wrapping_mul(quo_part) << extra_shl);
+                } else {
+                    // Two possibility algorithm
+                    let shift = n - duo_lz;
+                    let duo_sig_n = (duo >> shift) as $uX;
+                    let div_sig_n = (div >> shift) as $uX;
+                    let quo_part = $half_division(duo_sig_n, div_sig_n).0;
+                    let div_lo = div as $uX;
+                    let div_hi = (div >> n) as $uX;
+
+                    let (tmp_lo, carry) = carrying_mul(quo_part, div_lo);
+                    // The undersubtracting long division algorithm has already run once, so
+                    // overflow beyond `$uD` bits is not possible here
+                    let (tmp_hi, _) = carrying_mul_add(quo_part, div_hi, carry);
+                    let tmp = (tmp_lo as $uD) | ((tmp_hi as $uD) << n);
+
+                    if duo < tmp {
+                        return (
+                            quo + ((quo_part - 1) as $uD),
+                            duo.wrapping_add(div).wrapping_sub(tmp)
+                        )
+                    } else {
+                        return (
+                            quo + (quo_part as $uD),
+                            duo - tmp
+                        )
+                    }
+                }
+
+                duo_lz = duo.leading_zeros();
+
+                if div_lz <= duo_lz {
+                    // quotient can have 0 or 1 added to it
+                    if div <= duo {
+                        return (
+                            quo + 1,
+                            duo - div
+                        )
+                    } else {
+                        return (
+                            quo,
+                            duo
+                        )
+                    }
+                }
+
+                // This can only happen if `div_sd < n` (because of previous "quo = 0 or 1"
+                // branches), but it is not worth it to unroll further.
+                if n <= duo_lz {
+                    // simple division and addition
+                    let tmp = $half_division(duo as $uX, div as $uX);
+                    return (
+                        quo + (tmp.0 as $uD),
+                        tmp.1 as $uD
+                    )
+                }
+            }
+        }
+
+        /// Computes the quotient and remainder of `duo` divided by `div` and returns them as a
+        /// tuple.
+        ///
+        /// # Panics
+        ///
+        /// When attempting to divide by zero, this function will panic.
+        $(
+            #[$signed_attr]
+        )*
+        pub fn $signed_name(duo: $iD, div: $iD) -> ($iD, $iD) {
```

Comment:
I found this strategy to be the best one. In particular, it helps the compiler reduce the branches it takes to negate inputs when it knows that particular inputs will always be positive. On the current compiler, I found that a function:
```
pub fn lkj(lhs: i128, rhs: i128) -> i128 {
        lhs / rhs
}
```
would call `__divti3`. If I instead used:
```
pub fn lkj(lhs: i128, rhs: i128) -> i128 {
    if lhs > 0 && rhs > 0 {
        lhs / rhs
    } else {0}
}
```
the assembly shows that `__udivti3` is called directly.

Reply:
This is something that LLVM does as an optimisation in user code. In particular it will replace

```llvm
%3 = sdiv iN %0, %1 # signed division
```

with 

```
%7 = udiv iN %0, %1 # unsigned division
```

if it can tell %0 and %1 are both non-negative. The implementation of division in `compiler-builtins` should not have any effect for this optimizer behaviour.

---

I don’t particularly care much about what the implementation of this function is currently, they should all be pretty much equivalent in terms of speed. Just found it curious that it is repeated across different (trifecta/asymmetric/etc) strategies.